File: addBinary/addBinary.py
class Solution:
    def addBinary(self, a: 'str', b: 'str') -> 'str':
        lenA = len(a)-1
        lenB = len(b)-1
        result = ""
        c = 0
        while(lenA>=0 or lenB>=0 or c==1):
            aa = int(a[lenA]) if lenA>=0 else 0
            bb = int(b[lenB]) if lenB>=0 else 0
            result = str(aa^bb^c) + result
            c = 1 if aa+bb+c>=2 else 0
            lenA-=1
            lenB-=1
        return result

# 递归
class Solution:
    def addBinary(self, a: 'str', b: 'str') -> 'str':
        if len(a)==0: return b
        if len(b)==0: return a
        if int(a[-1])==1 and int(b[-1])==1:
            return self.addBinary(self.addBinary(a[0:-1],b[0:-1]),'1')+'0'
        elif int(a[-1])==0 and int(b[-1])==0:
            return self.addBinary(a[0:-1],b[0:-1])+'0'
        else:
            return self.addBinary(a[0:-1],b[0:-1])+'1'
            

File: addBinary/addBinary.java
class Solution {
    public String addBinary(String a, String b) {
        if(a.length()==0&&b.length()==0) return "0";
        StringBuilder result = new StringBuilder();
        int cc = 0;
        int i = a.length()-1;
        int j = b.length()-1;
        while(i>=0||j>=0||cc!=0){
            int add = 0;
            int addA = i>=0?(a.charAt(i)=='1'?1:0):0;
            int addB = j>=0?(b.charAt(j)=='1'?1:0):0;
            
            add = addA+addB+cc;
            System.out.print(add);
            if(add>=2){
                result.append(add==3?1:0);
                cc = 1;
            }else{
                result.append(add);
                cc = 0;
            }
            i--;
            j--;
        }
        return result.reverse().toString();
        
    }
}

class Solution {
    public String addBinary(String a, String b) {
        if(a.length()==0&&b.length()==0) return "0";
        StringBuilder result = new StringBuilder();
        int cc = 0;
        int i = a.length()-1;
        int j = b.length()-1;
        while(i>=0||j>=0||cc!=0){
            int add = 0;
            int addA = i>=0?(a.charAt(i)=='1'?1:0):0;
            int addB = j>=0?(b.charAt(j)=='1'?1:0):0;
            
            add = addA^addB^cc;
            cc = (addA+addB+cc)>=2?1:0;
            result.append(add);
            i--;
            j--;
        }
        return result.reverse().toString();
        
    }
}